### PR TITLE
Add inputs for sample tests

### PR DESCRIPTION
**Pull request by blindpirate**
### Context

Some tests depend on sample directories (for example this one https://github.com/gradle/gradle/blob/96eecdf0f6c6c942470678ef4e2ce325ac3851fa/subprojects/ivy/src/integTest/groovy/org/gradle/api/AutoTestedSamplesIvyIntegrationTest.groovy#L26) but didn't mark them as inputs. This breaks distributed test.

### CODE REVIEW

**Comment by blindpirate** [comment]
@bot-gradle test this

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/shared-configuration.kt [hunk 1]
```diff
@@ -90,14 +90,16 @@ fun Project.createTestTask(name: String, executer: String, sourceSet: SourceSet,
...
-
-fun splitIntoBuckets(sourceFiles: List<File>, numberOfSplits: Int): List<List<File>> =
-    if (sourceFiles.size % numberOfSplits == 0) {
-        sourceFiles.chunked(sourceFiles.size / numberOfSplits)
-    } else {
-        sourceFiles.chunked(sourceFiles.size / numberOfSplits + 1)
+/**
+ * Distributed test requires all dependencies to be declared
+ */
+fun Project.makeIntegrationTestsDependOnSampleDir(vararg sampleDirs: String) {
```
Comment (wolfs) [inline buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/shared-configuration.kt:96]:
```suggestion
fun Project.integrationTestUsesSampleDir(vararg sampleDirs: String) {
```

File: subprojects/ide/ide.gradle.kts [hunk 2]
```diff
@@ -82,3 +84,5 @@ classycle {
 testFilesCleanup {
     policy.set(WhenNotEmpty.REPORT)
 }
+
+makeIntegrationTestsDependOnSampleDir("subprojects/ide/src/main")
```
Comment (wolfs) [inline subprojects/ide/ide.gradle.kts:88]: Is there a way we can fail the build if the input is not declared?

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/shared-configuration.kt [hunk 3]
```diff
@@ -90,14 +90,16 @@ fun Project.createTestTask(name: String, executer: String, sourceSet: SourceSet,
...
+    tasks.withType<IntegrationTest>() {
+        inputs.files(sampleDirs.map { rootProject.fileTree(it) }.toTypedArray())
```
Comment (wolfs) [inline buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/shared-configuration.kt:98]:
```suggestion
        inputs.files(rootProject.files(it))
```

No need to use `fileTree`. Using `toTypedArray()` probably also resolves the `fileTree()`, which seems like a bad idea.

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/shared-configuration.kt [hunk 4]
```diff
@@ -90,14 +90,16 @@ fun Project.createTestTask(name: String, executer: String, sourceSet: SourceSet,
...
+            .withPropertyName("autoSampleTestFiles")
```
Comment (wolfs) [inline buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/shared-configuration.kt:99]:
```suggestion
            .withPropertyName("autoTestedSamples")
```

File: subprojects/ide/ide.gradle.kts (hunk 2 above)
Comment (blindpirate) [reply subprojects/ide/ide.gradle.kts:88]: We can set a Test system property when declaring the inputs, then assert that system property during sample discovering process. Then a new test without declaring inputs fails immediately.

**Comment by blindpirate** [comment]
@bot-gradle test QF

**Comment by blindpirate** [comment]
@bot-gradle test QF

**Comment by blindpirate** [comment]
@bot-gradle test QF

File: subprojects/ide/ide.gradle.kts (hunk 2 above)
Comment (blindpirate) [reply subprojects/ide/ide.gradle.kts]: It works immediately... https://builds.gradle.org/viewLog.html?buildId=34992820&tab=buildResultsDiv&buildTypeId=Gradle_Check_Quick_1_bucket22

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/shared-configuration.kt [hunk 5]
```diff
@@ -91,12 +91,17 @@ fun Project.createTestTask(name: String, executer: String, sourceSet: SourceSet,
...
-    if (sourceFiles.size % numberOfSplits == 0) {
-        sourceFiles.chunked(sourceFiles.size / numberOfSplits)
-    } else {
-        sourceFiles.chunked(sourceFiles.size / numberOfSplits + 1)
+/**
+ * Distributed test requires all dependencies to be declared
+ */
+fun Project.integrationTestUsesSampleDir(vararg sampleDirs: String) {
+    tasks.withType<IntegrationTest>() {
+        systemProperty("autoTestedSamplesDeclared", "true")
```
Comment (wolfs) [inline buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/shared-configuration.kt:99]: Should we be more specific which samples are declared?

File: subprojects/language-java/language-java.gradle.kts [hunk 6]
```diff
@@ -78,3 +79,6 @@ classycle {
 testFilesCleanup {
     policy.set(WhenNotEmpty.REPORT)
 }
+
+
```
Comment (wolfs) [inline subprojects/language-java/language-java.gradle.kts:83]: Nit: extra new line.

File: subprojects/language-native/language-native.gradle.kts [hunk 7]
```diff
@@ -81,3 +82,6 @@ dependencies {
 classycle {
     excludePatterns.set(listOf("org/gradle/language/nativeplatform/internal/**"))
 }
+
+
```
Comment (wolfs) [inline subprojects/language-native/language-native.gradle.kts:86]: Nit: extra new line.

**Comment by blindpirate** [comment]
@bot-gradle test QF

File: buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/shared-configuration.kt (hunk 5 above)
Comment (blindpirate) [reply buildSrc/subprojects/integration-testing/src/main/kotlin/org/gradle/gradlebuild/test/integrationtests/shared-configuration.kt:99]: Sure, yes.

**Comment by blindpirate** [comment]
@bot-gradle test QF

**Comment by blindpirate** [comment]
I need this change so I'm going to merge it, thanks @wolfs !

**Pull request merged**